aptos-move/block-executor/src/executor.rs

```diff
@@ -1262,22 +1191,29 @@ where
 
                     if dynamic_change_set_optimizations_enabled {
                         // Replace delayed field id with values in resource write set and read set.
-                        let delayed_field_keys =
-                            Some(output.delayed_field_change_set().into_keys());
                         let resource_change_set = Some(output.resource_write_set());
-                        let (mut patched_resource_write_set, write_set_keys) =
+                        let mut patched_resource_write_set =
                             Self::map_id_to_values_in_write_set(resource_change_set, &latest_view);
 
-                        let read_set = latest_view.read_set_sequential_execution();
-                        patched_resource_write_set.extend(
-                            Self::map_id_to_values_in_read_set_sequential(
-                                delayed_field_keys,
-                                write_set_keys,
-                                read_set,
-                                &unsync_map,
-                                &latest_view,
-                            ),
-                        );
+                        for (key, (value, layout)) in
+                            output.reads_needing_delayed_field_exchange().into_iter()
+                        {
+                            if patched_resource_write_set
+                                .insert(
+                                    key,
+                                    Self::replace_ids_with_values(
+                                        &value,
+                                        layout.as_ref(),
+                                        &latest_view,
+                                    ),
+                                )
+                                .is_some()
+                            {
+                                return Err(Error::FallbackToSequential(code_invariant_error(
```

Comment:
I didn't get this. Why is it a code invariant error if `patched_resource_write_set.insert(...)` returns `Some(..)`?

Reply:
we shouldn't be patching same resource twice. 

reads_needing_delayed_field_exchange shouldn't contain any resources that are already in the write set (filtering happens when creating VMChangeSet - and filtering is required for proper gas charging)